wasmbinding/message_plugin.go

```diff
@@ -193,6 +206,138 @@ func (m *CustomMessenger) submitTx(ctx sdk.Context, contractAddr sdk.AccAddress,
 	return nil, [][]byte{data}, nil
 }
 
+func (m *CustomMessenger) submitProposal(ctx sdk.Context, contractAddr sdk.AccAddress, submitProposal *bindings.SubmitProposal) ([]sdk.Event, [][]byte, error) {
+	response, err := m.PerformSubmitProposal(ctx, contractAddr, submitProposal)
+	if err != nil {
+		ctx.Logger().Debug("PerformSubmitTx: failed to submitProposal",
+			"from_address", contractAddr.String(),
+			"creator", contractAddr.String(),
+			"error", err,
+		)
+		return nil, nil, sdkerrors.Wrap(err, "failed to submit add admin message")
+	}
+
+	data, err := json.Marshal(response)
+	if err != nil {
+		ctx.Logger().Error("json.Marshal: failed to marshal submitProposal response to JSON",
+			"from_address", contractAddr.String(),
+			"creator", contractAddr.String(),
+			"error", err,
+		)
+		return nil, nil, sdkerrors.Wrap(err, "marshal json failed")
+	}
+
+	ctx.Logger().Debug("submit proposal message submitted",
+		"from_address", contractAddr.String(),
+		"creator", contractAddr.String(),
+	)
+	return nil, [][]byte{data}, nil
+}
+
+func (m *CustomMessenger) PerformSubmitProposal(ctx sdk.Context, contractAddr sdk.AccAddress, submitProposal *bindings.SubmitProposal) (*admintypes.MsgSubmitProposalResponse, error) {
+	msg := admintypes.MsgSubmitProposal{Proposer: contractAddr.String()}
+
+	if submitProposal.Proposals.TextProposal != nil {
+
+		prop := govtypes.TextProposal{
+			Title:       submitProposal.Proposals.TextProposal.Title,
+			Description: submitProposal.Proposals.TextProposal.Description,
+		}
+		cont, err := proto.Marshal(&prop)
+		if err != nil {
+			return nil, sdkerrors.Wrap(err, "failed to marshall incoming SubmitProposal message")
+		}
+		msg.Content = &types.Any{
+			TypeUrl: "/cosmos.gov.v1beta1.TextProposal",
+			Value:   cont,
+		}
+
+	} else if submitProposal.Proposals.ParamChangeProposal != nil {
+		proposal := submitProposal.Proposals.ParamChangeProposal
+		prop := paramChange.ParameterChangeProposal{
+			Title:       submitProposal.Proposals.TextProposal.Title,
+			Description: submitProposal.Proposals.TextProposal.Description,
+		}
+		var parameterChanges []paramChange.ParamChange
+		for _, parameterChange := range proposal.Changes {
```

Comment:
What about making `proposal.Changes` in `bindings.SubmitProposal` as `paramChange.ParamChange` type?
You don't need this loop then.